secureboot_policy.go

```diff
@@ -0,0 +1,1292 @@
+
+	kekFilename     = "KEK-8be4df61-93ca-11d2-aa0d-00e098032b8c"       // Filename in efivarfs for accessing the KEK database
+	dbFilename      = "db-d719b2cb-3d3a-4596-a3bc-dad00e67656f"        // Filename in efivarfs for accessing the EFI authorized signature database
+	dbxFilename     = "dbx-d719b2cb-3d3a-4596-a3bc-dad00e67656f"       // Filename in efivarfs for accessing the EFI forbidden signature database
+	mokListFilename = "MokListRT-605dab50-e046-4300-abb6-3dd810dd8b23" // Filename in efivarfs for accessing a runtime copy of the shim MOK database
+
+	uefiDriverPCR      = 2 // UEFI Drivers and UEFI Applications PCR
+	bootManagerCodePCR = 4 // Boot Manager Code and Boot Attempts PCR
+	secureBootPCR      = 7 // Secure Boot Policy Measurements PCR
+
+	returningFromEfiApplicationEvent = "Returning from EFI Application from Boot Option" // EV_EFI_ACTION index 2: "Attempt to execute code from Boot Option was unsuccessful"
+
+	sbKeySyncExe = "sbkeysync"
+
+	winCertTypePKCSSignedData uint16 = 0x0002 // WIN_CERT_TYPE_PKCS_SIGNED_DATA
+	winCertTypeEfiGuid        uint16 = 0x0EF1 // WIN_CERT_TYPE_EFI_GUID
+)
+
+var (
+	shimGuid                     = tcglog.NewEFIGUID(0x605dab50, 0xe046, 0x4300, 0xabb6, [...]uint8{0x3d, 0xd8, 0x10, 0xdd, 0x8b, 0x23}) // SHIM_LOCK_GUID
+	efiGlobalVariableGuid        = tcglog.NewEFIGUID(0x8be4df61, 0x93ca, 0x11d2, 0xaa0d, [...]uint8{0x00, 0xe0, 0x98, 0x03, 0x2b, 0x8c}) // EFI_GLOBAL_VARIABLE
+	efiImageSecurityDatabaseGuid = tcglog.NewEFIGUID(0xd719b2cb, 0x3d3a, 0x4596, 0xa3bc, [...]uint8{0xda, 0xd0, 0x0e, 0x67, 0x65, 0x6f}) // EFI_IMAGE_SECURITY_DATABASE_GUID
+
+	efiCertX509Guid      = tcglog.NewEFIGUID(0xa5c059a1, 0x94e4, 0x4aa7, 0x87b5, [...]uint8{0xab, 0x15, 0x5c, 0x2b, 0xf0, 0x72}) // EFI_CERT_X509_GUID
+	efiCertTypePkcs7Guid = tcglog.NewEFIGUID(0x4aafd29d, 0x68df, 0x49ee, 0x8aa9, [...]uint8{0x34, 0x7d, 0x37, 0x56, 0x65, 0xa7}) // EFI_CERT_TYPE_PKCS7_GUID
+
+	eventLogPath = "/sys/kernel/security/tpm0/binary_bios_measurements" // Path of the TCG event log for the default TPM, in binary form
+	efivarsPath  = "/sys/firmware/efi/efivars"                          // Default mount point for efivarfs
+)
+
+// EFIImage corresponds to a binary that is loaded, verified and executed before ExitBootServices.
+type EFIImage interface {
+	fmt.Stringer
+	ReadAll() ([]byte, error) // Read the entire contents of the image
+}
+
+// SnapFileEFIImage corresponds to a binary contained within a snap file that is loaded, verified and executed before ExitBootServices.
+type SnapFileEFIImage struct {
+	Container snap.Container
+	Path      string // The path of the snap image (used by the implementation of fmt.Stringer)
+	FileName  string // The filename within the snap squashfs
+}
+
+func (f SnapFileEFIImage) String() string {
+	return f.Path + ":" + f.FileName
+}
+
+func (f SnapFileEFIImage) ReadAll() ([]byte, error) {
```

Comment:
I'm not sure I understand how this API will be used, won't []byte be quite large in some cases to just get all of it into memory?

Reply:
do we need to do something on the snapd side to help with this?

Reply:
snapd never implemented anything but ReadFile on Container because we didn't have large file use cases to consider

Reply:
ReadAll() is certainly suboptimal - ideally, we'd have access to a io.ReaderAt to avoid having to read the entire file in to memory (which, in the case of the kernel image is over 10MB)

Reply:
I looked, we should be able to add something like RandomAccessFile() ReadAtCloser to Container,  where ReadAtCloser is basically  io.ReaderAt + io.Closer. I'll try to have a PR up today, shouldn't block this PR though.

Reply:
Yeah, that would work for this - thanks!

Reply:
I proposed this:

https://github.com/snapcore/snapd/pull/8248